Sources/src/FromSTMTestBeamData_source.cc

```diff
@@ -0,0 +1,335 @@
+
+    currentSubRunNumber_ = 0;
+    currentEventNumber_ = 0;
+
+    printAtEvent = 0;
+
+  }
+
+
+  //----------------------------------------------------------------
+  void STMTestBeamDataDetail::readFile(const std::string& filename, art::FileBlock*& fb) {
+
+    // Open the input file
+    currentFileName_ = filename;
+    currentFile_ = new std::ifstream(currentFileName_.c_str(), std::ios::in | std::ios::binary);
+    if (!currentFile_->is_open()) {
+      throw cet::exception("FromSTMTestBeamData") << "A problem opening binary file " << currentFileName_ << std::endl;
+    }
+
+    // Extract the run number and subrun number from the file name
+    std::string delimeter = ".";
+    size_t currentPos = 0;
+    size_t previousPos = currentPos;
+    std::vector<std::string> tokens;
+    while ( (currentPos = currentFileName_.find(delimeter, previousPos)) != std::string::npos) {
+      tokens.push_back(currentFileName_.substr(previousPos, currentPos-previousPos));
+      previousPos = currentPos+1;
+    }
+    tokens.push_back(currentFileName_.substr(previousPos, currentPos-previousPos)); // add the final token which gets missed in the above loop
+    unsigned int n_expected_fields = 6;
+    if (tokens.size() != n_expected_fields) {
+      throw cet::exception("FromSTMTestBeamData") << "Number of fields in filename (" << tokens.size() << ") is not the same number we expected (" << n_expected_fields << "). Filename = " << currentFileName_ << std::endl;
+    }
+
+    // Get the channel from the configuration field
+    std::string configuration = tokens.at(3);
+    if (configuration.find("HPGe") != std::string::npos) {
+      channel_ = STMChannel::HPGe;
+    }
+    else if (configuration.find("LaBr") != std::string::npos) {
+      channel_ = STMChannel::LaBr;
+    }
+    else {
+      throw cet::exception("FromSTMTestBeamData") << "Cannot determine the channel from the configuration field (" << configuration << "). This should contain either \"HPGe\" or \"LaBr\"" << std::endl;
+    }
+
+    // Get the run and subrun numbers from the sequencer
+    std::string sequencer = tokens.at(4);
+    std::string runNo = sequencer.substr(0, 6); // sequencer always has 6 characters for run
+    runNumber_ = std::stoi(runNo);
+
+    if(!art::RunID(runNumber_).isValid()) {
+      throw cet::exception("BADCONFIG", " FromSTMTestBeamData: ")
+        << " fhicl::ParameterSet specifies an invalid runNumber = "<<runNumber_<<"\n";
+    }
+
+    if(runNumber_ < 101000 || runNumber_ > 101999) {
+      throw cet::exception("FromSTMTestBeamData") << "Run number is outside of our reserved run range (101000 -- 101999)" << std::endl;
+    }
+
+    // There are two different versions of the binary file:
+    //  - v2 contains a unix timestamp after the trigger header
+    //  - v1 does not
+    // For the time being, just hardcode which runs belong to which (ideally would be in a DB?)
+    if (runNumber_ == 101001) {
```

Comment:
Also here, OK for the test beam run, but not permanently, IMO.  Ideally the data would contain the version number and the code can react to that.  This is a general problem in all the data taking, but I don't know of a general rule for a solution.

Reply:
If you backed in a corner and need to test on run numbers, I suggest writing a small class or function to manage the tests; give the functions good names so that people can just read the code; and document that class/function well. This captures all of the hard coded run numbers in one place.

Reply:
I agree that this is not a permanent solution. In the future we will hopefully be reading art files as our inputs and not our own binary files (which had a unix time stamp added during the test beam)